File: src/voyage_geo/stages/query_generation/strategies/competitor.py
"""Competitor strategy — queries about alternatives and head-to-head comparisons."""

from __future__ import annotations

from datetime import UTC, datetime

from voyage_geo.providers.base import BaseProvider
from voyage_geo.stages.query_generation.strategies.parse import parse_ai_queries
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.query import GeneratedQuery

YEAR = datetime.now(UTC).year


def _build_prompt(profile: BrandProfile, count: int) -> str:
    competitors = ", ".join(profile.competitors) if profile.competitors else "major alternatives in the space"
    return f"""Generate {count} search queries about the competitive landscape in "{profile.category}".

CONTEXT:
- Category: {profile.category}
- Industry: {profile.industry}
- Known competitors: {competitors}
- Year: {YEAR}

These queries CAN and SHOULD mention competitor names from the list above. The goal is to see if AI models recommend {profile.name} when someone asks about competitors.

EXAMPLES of realistic competitor queries (for "spend management platform", competitors: Ramp, Divvy, Airbase):
- ramp alternatives {YEAR} | alternative | switching
- divvy vs other corporate card options | comparison | head-to-head
- what are the best alternatives to airbase | alternative | discovery
- companies like ramp but cheaper | alternative | pricing
- is there something better than divvy for startups | alternative | evaluation
- top corporate card platforms besides ramp | best-of | landscape

YOUR QUERIES MUST BE:
- SHORT: 5-15 words max
- NATURAL: how real people ask chatbots for competitor info
- Use competitor names from the list — rotate through them

RULES:
- NEVER mention {profile.name} — that's the brand we're measuring
- Keep queries short and direct
- Each query on its own line:

<query text> | <category> | <intent>

Categories: comparison, alternative, recommendation, review, best-of, general
Intent: head-to-head, switching, discovery, pricing, evaluation, landscape

Generate exactly {count} queries:"""


async def generate(profile: BrandProfile, count: int, provider: BaseProvider) -> list[GeneratedQuery]:
    prompt = _build_prompt(profile, count)
    response = await provider.query(prompt)
    return parse_ai_queries(response.text, "competitor", "cp", count)

File: src/stages/query-generation/strategies/competitor.ts
import type { BrandProfile } from '../../../types/brand.js';
import type { GeneratedQuery } from '../../../types/query.js';
import type { AIProvider } from '../../../providers/types.js';
import type { QueryStrategyInterface } from './types.js';
import { parseAIQueries } from './parse.js';

const YEAR = new Date().getFullYear();

const buildPrompt = (profile: BrandProfile, count: number): string => {
  const competitors = profile.competitors.length > 0
    ? profile.competitors.join(', ')
    : 'major alternatives in the space';

  return `You are a GEO (Generative Engine Optimization) specialist. Generate ${count} realistic AI chatbot queries focused on COMPETITIVE COMPARISONS in the "${profile.category}" space.

BRAND CONTEXT:
- Brand: ${profile.name}
- Category: ${profile.category}
- Industry: ${profile.industry}
- Known competitors: ${competitors}
- Keywords: ${profile.keywords.join(', ') || profile.category}
- Year: ${YEAR}

QUERY TYPES TO MIX:
- Head-to-head: "${profile.name} vs [competitor] — which is better?"
- Decision queries: "Should I use X or Y?", "Trying to decide between X and Y"
- Switching: "Thinking of switching from [competitor] to ${profile.name}", "Is [competitor] still the best or should I look at alternatives?"
- Alternative-seeking: "[competitor] alternatives", "Something better than [competitor]?"
- Value comparison: "Which has better pricing, X or Y?", "X vs Y for the money"
- Use-case comparison: "X vs Y for [specific use case from keywords]"
- Justification: "Why would someone choose X over Y?", "What does X do that Y doesn't?"
- Team/scale: "X vs Y for large teams", "Which scales better?"

RULES:
- Use the ACTUAL competitor names listed above — rotate through them
- Queries should sound like real people asking AI chatbots for help deciding
- Mix ${profile.name} appearing first and second in comparisons
- Some queries should not mention ${profile.name} at all — just compare competitors or ask for alternatives
- Conversational, natural phrasing — not templated
- Include ${YEAR} naturally in a few queries
- Do NOT use placeholder brackets

FORMAT (one per line):
<query text> | <category> | <intent>

Categories: comparison, alternative, recommendation, review, general
Intent: head-to-head, decision, switching, migration, evaluation, cost, use-case, justification, trust, team

Generate exactly ${count} queries now:`;
};

export class CompetitorStrategy implements QueryStrategyInterface {
  name = 'competitor';

  async generate(profile: BrandProfile, count: number, provider: AIProvider): Promise<GeneratedQuery[]> {
    const prompt = buildPrompt(profile, count);
    const response = await provider.query(prompt);
    return parseAIQueries(response.text, 'competitor', 'cp', count);
  }
}
